lib/Sema/DerivedConformanceComparable.cpp

```diff
@@ -0,0 +1,411 @@
+    param->setSpecifier(ParamSpecifier::Default);
+    param->setInterfaceType(selfIfaceTy);
+    return param;
+  };
+
+  ParameterList *params = ParameterList::create(C, {
+    getParamDecl("a"),
+    getParamDecl("b")
+  });
+
+  auto boolTy = C.getBoolDecl()->getDeclaredType();
+
+  Identifier generatedIdentifier;
+  if (parentDC->getParentModule()->isResilient()) {
+    generatedIdentifier = C.Id_LessThanOperator;
+  } else {
+    assert(selfIfaceTy->getEnumOrBoundGenericEnum());
+    generatedIdentifier = C.Id_derived_enum_less_than;
+  }
+
+  DeclName name(C, generatedIdentifier, params);
+  auto comparableDecl =
+    FuncDecl::create(C, /*StaticLoc=*/SourceLoc(),
+                     StaticSpellingKind::KeywordStatic,
+                     /*FuncLoc=*/SourceLoc(), name, /*NameLoc=*/SourceLoc(),
+                     /*Throws=*/false, /*ThrowsLoc=*/SourceLoc(),
+                     /*GenericParams=*/nullptr,
+                     params,
+                     TypeLoc::withoutLoc(boolTy),
+                     parentDC);
+  comparableDecl->setImplicit();
+  comparableDecl->setUserAccessible(false);
+
+  // Add the @_implements(Comparable, < (_:_:)) attribute
+  if (generatedIdentifier != C.Id_LessThanOperator) {
+    auto comparable = C.getProtocol(KnownProtocolKind::Comparable);
+    auto comparableType = comparable->getDeclaredType();
+    auto comparableTypeLoc = TypeLoc::withoutLoc(comparableType);
+    SmallVector<Identifier, 2> argumentLabels = { Identifier(), Identifier() };
+    auto comparableDeclName = DeclName(C, DeclBaseName(C.Id_LessThanOperator),
+                                   argumentLabels);
+    comparableDecl->getAttrs().add(new (C) ImplementsAttr(SourceLoc(),
+                                                          SourceRange(),
+                                                          comparableTypeLoc,
+                                                          comparableDeclName,
+                                                          DeclNameLoc()));
+  }
+
+  if (!C.getLessThanIntDecl()) {
+    derived.ConformanceDecl->diagnose(diag::no_less_than_overload_for_int);
+    return nullptr;
+  }
+
+  comparableDecl->setBodySynthesizer(bodySynthesizer);
+
+  comparableDecl->copyFormalAccessFrom(derived.Nominal, /*sourceIsParentContext*/ true);
+
+  // Add the operator to the parent scope.
+  derived.addMembersToConformanceContext({comparableDecl});
+
+  return comparableDecl;
+}
+
+bool 
+DerivedConformance::canDeriveComparable(DeclContext *context, NominalTypeDecl *declaration) {
+  // The type must be an enum.
+  if (EnumDecl *const enumeration = dyn_cast<EnumDecl>(declaration)) {
+    // The cases must not have non-comparable associated values or raw backing
+    auto comparable = context->getASTContext().getProtocol(KnownProtocolKind::Comparable);
+    return allAssociatedValuesConformToProtocol(context, enumeration, comparable) && !enumeration->hasRawType();
+  } else {
+    return false;
```

Comment:
Nit: Early-exit would undo a level of block structure here.

Reply:
done